libc/dirent/opendir.c

```diff
@@ -0,0 +1,443 @@
+
+	/* No execute permission */
+	chmod(unreadable, 0600);
+	errno = 0;
+	dirPtr = opendir(readable);
+	TEST_ASSERT_EQUAL_INT(EACCES, errno);
+	TEST_ASSERT_NULL(dirPtr);
+
+	/* No read permission */
+	chmod(unreadable, 0300);
+	errno = 0;
+	dirPtr = opendir(unreadable);
+	TEST_ASSERT_EQUAL_INT(EACCES, errno);
+	TEST_ASSERT_NULL(dirPtr);
+
+
+	errno = 0;
+	chmod(unreadable, 0700);
+	rmdir(readable);
+	chmod(unreadable, 0000);
+}
+
+
+TEST(opendir, wrong_directory_name)
+{
+	errno = 0;
+	DIR *dirPtr = opendir(MAIN_DIR "/not_existing_directory");
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+	TEST_ASSERT_NULL(dirPtr);
+
+	errno = 0;
+	dirPtr = opendir("");
+	TEST_ASSERT_NULL(dirPtr);
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+}
+
+
+TEST(opendir, not_a_directory)
+{
+	errno = 0;
+	DIR *dirPtr = opendir(MAIN_DIR "/notadir.txt");
+	TEST_ASSERT_EQUAL_INT(ENOTDIR, errno);
+	TEST_ASSERT_NULL(dirPtr);
+}
+
+
+TEST(opendir, creating_dirs_in_closed_and_open_directories)
+{
+	/* Create dir in closed directory */
+	mkdir(MAIN_DIR "/formerDir", 0777);
+	DIR *dirs[4];
+	/* Create dir in opened directory, then close opened one */
+	DIR *dp = opendir(MAIN_DIR);
+	TEST_ASSERT_NOT_NULL(dirs[0] = opendir(MAIN_DIR "/formerDir"));
+
+	mkdir(MAIN_DIR "/latterDir", 0777);
+
+	closedir(dp);
+
+	/* Assure that both dirs can be opened without problems */
+	TEST_ASSERT_NOT_NULL(dirs[1] = opendir(MAIN_DIR "/formerDir"));
+	TEST_ASSERT_NOT_NULL(dirs[2] = opendir(MAIN_DIR "/latterDir"));
+	dp = opendir(MAIN_DIR);
+	closedir(dp);
+	mkdir(MAIN_DIR "/evenLatterDir", 0777);
+	TEST_ASSERT_NOT_NULL(dirs[3] = opendir(MAIN_DIR "/evenLatterDir"));
+
+	for (int i = 0; i < 4; ++i)
+		closedir(dirs[i]);
+
+	rmdir(MAIN_DIR "/formerDir");
+	rmdir(MAIN_DIR "/latterDir");
+	rmdir(MAIN_DIR "/evenLatterDir");
+}
+
+
+TEST(opendir, open_too_many_directories)
+{
+#ifdef OPEN_MAX
+	int dir_amount = OPEN_MAX + 100;
+	errno = 0;
+
+	TEST_ASSERT_EQUAL(-1, create_directories(dir_amount, 'd'));
+	TEST_ASSERT_EQUAL(EMFILE, errno);
+#else
+	TEST_IGNORE_MESSAGE("OPEN_MAX not defined");
+#endif
+}
+
+
+TEST(opendir, open_small_enough_number_of_directories)
+{
+#ifdef OPEN_MAX
+	errno = 0;
+
+	TEST_ASSERT_EQUAL(0, create_directories(OPEN_MAX - 20, 'c'));
+#else
+	TEST_IGNORE_MESSAGE("OPEN_MAX not defined");
+#endif
+}
+
+
+TEST(opendir, open_same_dir_multiple_times)
+{
+	DIR *dp1, *dp2, *dp3;
+
+	dp1 = opendir(MAIN_DIR);
+	dp2 = opendir(MAIN_DIR);
+	dp3 = opendir(MAIN_DIR);
+
+	TEST_ASSERT_NOT_EQUAL(dp1, dp2);
+	TEST_ASSERT_NOT_EQUAL(dp2, dp3);
+	TEST_ASSERT_NOT_EQUAL(dp1, dp3);
+
+	closedir(dp1);
+	closedir(dp2);
+	closedir(dp3);
+}
+
+
+TEST(opendir, symlink_loop)
+{
+#ifndef SYMLOOP_MAX
+	TEST_IGNORE_MESSAGE("SYMLOOP_MAX not defined");
+#else
+	mkdir("A", 0777);
+
+	symlink("A", "D");
+	symlink(".", "A/D");
+
+	char loopPath[SYMLOOP_MAX * 4 + 16];
+	strcpy(loopPath, "A/");
+
+	/* Create a path to barely valid symloop */
+	for (int i = 0; i < SYMLOOP_MAX - 2; ++i) {
+		strcat(loopPath, "D/D/");
+	}
+
```

Comment:
So, it doesn't work because an ELOOP error occurs. This is caused by the fact that you are creating nearly twice as many symlinks as the maximum allowed, with each D being a symlink to A.